flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackend.java

```diff
@@ -791,20 +784,215 @@ private void restoreInstance(
 		}
 
 		/**
-		 * Recovery from local incremental state.
+		 * Recovery from multi incremental states.
+		 * In case of rescaling, this method creates a temporary RocksDB instance for a key-groups shard. All contents
+		 * from the temporary instance are copied into the real restore instance and then the temporary instance is
+		 * discarded.
 		 */
-		private void restoreInstance(IncrementalLocalKeyedStateHandle localKeyedStateHandle) throws Exception {
+		void restoreFromMultiHandles(Collection<KeyedStateHandle> restoreStateHandles) throws Exception {
+
+			KeyGroupRange targetKeyGroupRange = stateBackend.keyGroupRange;
+
+			chooseTheBestStateHandleToInit(restoreStateHandles, targetKeyGroupRange);
+
+			int targetStartKeyGroup = stateBackend.getKeyGroupRange().getStartKeyGroup();
+			byte[] targetStartKeyGroupPrefixBytes = new byte[stateBackend.keyGroupPrefixBytes];
+			for (int j = 0; j < stateBackend.keyGroupPrefixBytes; ++j) {
+				targetStartKeyGroupPrefixBytes[j] = (byte) (targetStartKeyGroup >>> ((stateBackend.keyGroupPrefixBytes - j - 1) * Byte.SIZE));
+			}
+
+			for (KeyedStateHandle rawStateHandle : restoreStateHandles) {
+
+				if (!(rawStateHandle instanceof IncrementalKeyedStateHandle)) {
+					throw new IllegalStateException("Unexpected state handle type, " +
+						"expected " + IncrementalKeyedStateHandle.class +
+						", but found " + rawStateHandle.getClass());
+				}
+
+				Path temporaryRestoreInstancePath = new Path(stateBackend.instanceBasePath.getAbsolutePath() + UUID.randomUUID().toString());
+				try (RestoredDBInfo tmpRestoreDBInfo = restoreDBFromStateHandle(
+						(IncrementalKeyedStateHandle) rawStateHandle,
+						temporaryRestoreInstancePath,
+						targetKeyGroupRange,
+						stateBackend.keyGroupPrefixBytes)) {
+
+					List<ColumnFamilyDescriptor> tmpColumnFamilyDescriptors = tmpRestoreDBInfo.columnFamilyDescriptors;
+					List<ColumnFamilyHandle> tmpColumnFamilyHandles = tmpRestoreDBInfo.columnFamilyHandles;
+
+					// iterating only the requested descriptors automatically skips the default column family handle
+					for (int i = 0; i < tmpColumnFamilyDescriptors.size(); ++i) {
+						ColumnFamilyHandle tmpColumnFamilyHandle = tmpColumnFamilyHandles.get(i);
+						ColumnFamilyDescriptor tmpColumnFamilyDescriptor = tmpColumnFamilyDescriptors.get(i);
+
+						ColumnFamilyHandle targetColumnFamilyHandle = getOrRegisterColumnFamilyHandle(
+							tmpColumnFamilyDescriptor, null, tmpRestoreDBInfo.stateMetaInfoSnapshots.get(i));
+
+						try (RocksIterator iterator = tmpRestoreDBInfo.db.newIterator(tmpColumnFamilyHandle)) {
+
+							iterator.seek(targetStartKeyGroupPrefixBytes);
```

Comment:
If the DB is clipped, do we even need to seek or will the iterator already begin at a useful key-group anyways?

Reply:
I think this is a not so nice API of `RocksIterator`, a newly created Iterator doesn't point to the header element by default, users need to performance the `seek()` to make sure it is valid.